### PR TITLE
WIP, but working: Put email sending in a different process

### PR DESCRIPTION
**Pull request by binary1230**
HEY! I think this actually works!

This takes the email sending process off the main process and puts it in its own process, where it can successfully take advantage of multi-processing (since python threads can't execute at the same time).  To do this, I had to replace the global 'last email sending run' results object with DB tables that communicate the same info.  The email sending process will dump its last result into the DB now, and other processes/threads can consume it (such as the pending emails page, or the pending emails report that goes out).

Building on the scheduler foundation, this also means any arbitrary background task can be put on any arbitrary process.  They can be on the same box, or different boxes, or whatever.

This email sending code is triggered here via a 'sep' command.  So it's expected that some process (in our case, Supervisor) will startup and maintain the following two processes:
```
run_server .  # run sideboard web server
sep start_email_sending_daemon .  # run our email daemon
```

For safety, the code/config/db/etc would now need to be kept in sync at all times.

NOTE: this code is not off MASTER but instead off the scheduling rewrite branch #2976 , so be sure to check that one out too.

Needs the following to ship:
- [ ] way more testing, in particular more than just like, one email category on my dev server here
- [ ] DB migrations created (Easy)
- [ ] write new unit tests, probably
- [ ] fix existing unit tests
- [ ] merge #2976 and associated PR's first, which this code depends on
- [x] probably go just one more notch overboard, and still put the email sending in its own thread.  This is because all tasks are now executed serially via the scheduler, and if the email sending stays active a very long time, it could block out the other smaller reports/tasks from running in a timely fashion (used to take upwards of 10 minutes per run at a full-scale mag database)

### CODE REVIEW

**Comment by binary1230** [comment]
see https://github.com/magfest/ubersystem/issues/2978 for some more discussion

File: uber/automated_emails_server.py [hunk 1]
```diff
@@ -389,43 +388,45 @@ def get_pending_email_data():
...
-            'subject': automated_email.subject,
-            'sender': automated_email.sender,
-        }
 
     return pending_emails_by_sender
 
 
 uber.scheduler.register_task(
-    lambda: uber.scheduler.schedule.every(5).minutes.do(SendAllAutomatedEmailsJob.send_all_emails),
+    lambda: uber.scheduler.schedule.every(30).seconds.do(SendAllAutomatedEmailsJob.send_all_emails),
```
Comment (binary1230) [inline uber/automated_emails_server.py]: since the scheduler (and the email daemon itself) won't let overlapping jobs happen, no reason to not just shorten this window.

File: uber/automated_emails_server.py [hunk 2]
```diff
@@ -389,43 +388,45 @@ def get_pending_email_data():
...
     category="automated_email_sending",
 )
 
 
 @entry_point
 def start_email_sending_daemon():
+    initialize_db()
```
Comment (binary1230) [inline uber/automated_emails_server.py]: see #2980 about this, I'm not sure this is what we want to do.  seems to work ok though.

File: uber/decorators.py [hunk 3]
```diff
@@ -374,14 +374,17 @@ def with_caching(*args, **kwargs):
...
     def with_timing(*args, **kwargs):
         before = datetime.now()
         try:
             return func(*args, **kwargs)
         finally:
-            log.debug('{}.{} loaded in {} seconds'.format(func.__module__, func.__name__, (datetime.now() - before).total_seconds()))
+            prepend = prepend_txt
+            if prepend:
+                prepend += ': '
+            log.debug('{}{}.{} loaded in {} seconds'.format(prepend, func.__module__, func.__name__, (datetime.now() - before).total_seconds()))
```
Comment (binary1230) [inline uber/decorators.py]: unrelated change to just make the logging easier to understand

File: uber/models/email.py [hunk 4]
```diff
@@ -64,3 +65,27 @@ def html(self):
...
+        last_run = EmailDaemonStatus.last_result(session)
+        if not last_run:
+            return None
+
+        return last_run if last_run.last_run_finished > datetime.now(UTC) - timedelta(days=1) else None
+
+
+class EmailDaemonCategoryResult(MagModel):
+    ident = Column(UnicodeText)
+    unsent_because_unapproved = Column(Integer, default=0)
```
Comment (binary1230) [inline uber/models/email.py]: I specifically didn't try linking up foreign keys to the AutomatedEmails category here because I wanted this data to be a bit less tightly coupled and more like reading a report.  If keys/idents change, I wanted to make the client code tolerant of those changes without having to worry about it too much.

File: uber/scheduler.py [hunk 5]
```diff
@@ -8,14 +8,13 @@
 # don't start the email sending in the main thread. everything else is cool.
 # NOTE: only applies to cherrypy process ONLY, doesn't apply when started via 'sep' or other processes.
 # uncomment this if you want to do email sending in its own process
-# _default_cherrypy_exclude_categories = ["automated_email_sending"]
-_default_cherrypy_exclude_categories = None
+_default_cherrypy_exclude_categories = ["automated_email_sending"]
```
Comment (binary1230) [inline uber/scheduler.py]: this is the magic.  main process won't start the automated_email_sending category, but we will explicitly do it in the sideboard entry point

**Comment by RobRuana** [comment]
I am super excited about this! I'll note (and you already mentioned this) we'll need to update our deployment scripts to understand the logical difference between an uber web server and an uber email server (even if they live on the same box).

Once we fully de-thread our DaemonTasks, we'll probably also want our deployments to be a little smarter about how many uber web servers to run per-server. Probably one uber web server per core?

File: uber/scheduler.py [hunk 6]
```diff
@@ -1,21 +1,40 @@
...
 
 class UberSchedulerJob(schedule.Job):
-    def do(self, job_func, *args, **kwargs):
-        # our scheduler needs to not let individual tasks throw exceptions, so wrap it here
-        wrapped_fn = timed(swallow_exceptions(job_func))
+    def do(self, job_func, run_in_own_thread=False, thread_name=None, *args, **kwargs):
+        # our scheduler needs to not let individual tasks throw exceptions, so wrap it here, and time it
+        wrapped_fn = timed(swallow_exceptions(job_func), prepend_txt='TASK DONE')
+
+        if run_in_own_thread:
```
Comment (EliAndrewC) [inline uber/scheduler.py]: As a naming suggestion, I recommend changing this to ``run_in_new_thread`` because I originally misinterpreted "run in own thread" to mean "run in the owner's thread and NOT in a new thread".  Which is clearly wrong and kind of a dumb interpretation on my part, since presumably if you meant that you would have said ``run_in_caller_thread`` or something, but I was still confused.

File: uber/site_sections/emails.py [hunk 7]
```diff
@@ -21,26 +21,31 @@ def sent(self, session, **params):
...
         sent_email_counts = {c[0]: c[1] for c in count_query.all()}
 
         for automated_email in AutomatedEmail.instances.values():
-            category_results = categories_results.get(automated_email.ident, None) if categories_results else None
-            unsent_because_unapproved = category_results.get('unsent_because_unapproved', 0) if category_results else 0
+            ident = automated_email.ident
+            unsent_because_unapproved = 0
+
+            if categories_results:
+                category_results = next((result for result in categories_results if result.ident == ident), None)
```
Comment (EliAndrewC) [inline uber/site_sections/emails.py]: I recommend rewriting this line since it's somewhat confusing.  It took me a couple of re-reads to realize you were using a generator comprehension and using that to check for the presence of a result.

I recommend replacing this line and the original

```python
categories_results = session.query(EmailDaemonCategoryResult).all()
```

line with the following:

```python
category_results = session.query(EmailDaemonCategoryResult).filter_by(ident=ident).first()
```

which I'm pretty sure will do the same thing, unless I've misunderstood this code.

File: uber/models/email.py (hunk 4 above)
Comment (EliAndrewC) [reply uber/models/email.py]: I agree that the lack of coupling is a good idea in this case.  It'll probably make future refactors easier this way.

**Comment by RobRuana** [comment]
Superseded by https://github.com/magfest/ubersystem/pull/3164
